packages/dds/migration-shim/src/test/reconnect.spec.ts

```diff
@@ -0,0 +1,277 @@
+	// V1 of the code: Registry setup to create the old document
+	const oldChannelFactory = LegacySharedTree.getFactory();
+	const dataObjectFactory1 = new DataObjectFactory(
+		"TestDataObject",
+		TestDataObject,
+		[oldChannelFactory],
+		{},
+	);
+
+	// The 1st runtime factory, V1 of the code
+	const runtimeFactory1 = new ContainerRuntimeFactoryWithDefaultDataStore({
+		defaultFactory: dataObjectFactory1,
+		registryEntries: [dataObjectFactory1.registryEntry],
+	});
+
+	// V2 of the registry (the migration registry) -----------------------------------------
+	// V2 of the code: Registry setup to migrate the document
+	const legacySharedTreeFactory = LegacySharedTree.getFactory();
+	const newSharedTreeFactory = new SharedTreeFactory();
+
+	const migrationShimFactory = new MigrationShimFactory(
+		legacySharedTreeFactory,
+		newSharedTreeFactory,
+		(legacyTree, newTree) => {
+			// Migration code that the customer writes
+			// Revert local edits - otherwise we will be eventually inconsistent
+			const edits = legacyTree.edits as EditLog;
+			const localEdits = [...edits.getLocalEdits()].reverse();
+			for (const edit of localEdits) {
+				legacyTree.revert(edit.id);
+			}
+			// migrate data
+			const quantity = getQuantity(legacyTree);
+			newTree.schematize({
+				initialTree: {
+					quantity,
+				},
+				allowedSchemaModifications: AllowedUpdateType.None,
+				schema,
+			});
+		},
+	);
+
+	const sharedTreeShimFactory = new SharedTreeShimFactory(newSharedTreeFactory);
+
+	const dataObjectFactory2 = new DataObjectFactory(
+		"TestDataObject",
+		TestDataObject,
+		[migrationShimFactory, sharedTreeShimFactory], // Use the migrationShimFactory instead of the LegacySharedTreeFactory
+		{},
+	);
+
+	// The 2nd runtime factory, V2 of the code
+	const runtimeFactory2 = new ContainerRuntimeFactoryWithDefaultDataStore({
+		defaultFactory: dataObjectFactory2,
+		registryEntries: [dataObjectFactory2.registryEntry],
+		runtimeOptions,
+	});
+
+	let provider: ITestObjectProvider;
+
+	const originalValue = 3;
+	const newValue = 4;
+
+	beforeEach(async () => {
+		provider = getTestObjectProvider();
+		// Creates the document as v1 of the code with a SharedCell
+		const container = await provider.createContainer(runtimeFactory1);
+		const testObj = await requestFluidObject<TestDataObject>(container, "/");
+		const legacyTree = testObj.getTree<LegacySharedTree>();
+
+		updateQuantity(legacyTree, originalValue);
+		// make sure changes are saved.
+		await provider.ensureSynchronized();
+		container.close();
+	});
+
+	it("Shims can reconnect", async () => {
+		// Setup containers and get Migration Shims instead of LegacySharedTrees
+		const container1 = await provider.loadContainer(runtimeFactory2);
+		const testObj1 = await requestFluidObject<TestDataObject>(container1, "/");
+		const shim1 = testObj1.getTree<MigrationShim>();
+		const legacyTree1 = shim1.currentTree as LegacySharedTree;
+		container1.disconnect();
+		container1.connect();
```

Comment:
It's probably interesting to explicitly test "resubmit (of some legacy op) on reconnect (after migration has happened)" too?

Reply:
Yeah, I want to create separate tests for that. My understanding is that if we call resubmit on some legacy op, the new shared tree will blow up.